### DISCUSSION TITLE
Issue with sqlalchemy.text(): The column referenced alias is not updating with the nested queries

### DISCUSSION THREAD

**Discussion by SunilMallireddy** [sqlalchemy/sqlalchemy#9483] [Usage Questions]
Hi,

The problem is that when using sqlalchemy to create a new datetime column in BigQuery by passing an Interval expression to the date_add function using sa.text(), the query alias for the integer column is incorrect in the subsequent queries. The alias for the last_active column is updated correctly to "anon_2", but the integer column points to "anon_1" instead of "anon_2" in the query that uses the new datetime column. This causes the query to fail. The issue is that there does not seem to be an alternate date_add function in BigQuery that does not require an Interval expression, which would avoid the need to use sa.text(). The question is whether there is an alternative sqlalchemy function that can be used instead of sa.text().

Description of the below code:
I am trying to execute the following queries in the bigquery using sqlalchemy. I have created a new table "activity" having two columns "last_active" and "integer". Then I have created a query q1 using the two columns last_active and integer. On top of the query q1, I have created a new datetime column using the columns of q1. For creation of new datetime column, to pass the Interval expression in the date_add function I have used sa.text("Interval column_name days"). I have created a new query q2 using new datetime column. But, when I try to create a new datetime column using the q2 columns, or fetch the count of q2, there is an issue.
The query alias has been updated to "anon_2" for last_active column, which is correct, but where as for the integer column it points to "anon_1" query alias , which is incorrect and which is the query alias being used in q2, instead of "anon_2", which is correct alias to be referenced. So the query is failing as it pointing to the incorrect alias.

Issue: the issue is to pass the Interval expression to create a new date time column, I am using sqlalchemy.text() which is causing this issue. I cannot find the alternate date_add function is bigquery which doesn't use Interval expression and I have to pass Interval expression in my case. Is there any sqlalchemy function which can be used of instead of sa.text().

Note: These queries are being executed in the bigquery 

```python

import sqlalchemy as sa
##create a new table
table = sa.table('activity', sa.column('last_active', sa.DateTime()), sa.column('integer', sa.Integer()))
integer = table.columns['integer']
last_active = table.columns['last_active']

##create a subquery1
q1 = sa.select(last_active,integer).select_from(table)

## query = 'SELECT activity.last_active, activity.integer \nFROM activity'
q1_date_column = q1.columns['last_active']
q1_integer = q1.columns['integer']
new_col = sa.func.date_add(q1_date_column, sa.text('INTERVAL {q1_integer} days')).label('newdatetimecolumn')

#create a subquery2
q2 = sa.select(new_col)

#query = 'SELECT date_add(anon_1.last_active, INTERVAL anon_1.integer days) AS newdatetimecolumn \nFROM (SELECT activity.last_active AS last_active, activity.integer AS integer \nFROM activity) AS anon_1'

#create a new query
q3 = sa.select([sa.func.count(sa.text('*'))]).select_from(q2)

#query = str(q3)
##query = 'SELECT count(*) AS count_1 \nFROM (SELECT date_add(anon_2.last_active, INTERVAL anon_1.integer days) AS newdatetimecolumn \nFROM (SELECT activity.last_active AS last_active, activity.integer AS integer \nFROM activity) AS anon_2) AS anon_1'
```

**Comment by zzzeek** [comment]
havent run your example but important note, this is not a subquery:

    q1 = sa.select(last_active,integer).select_from(table)

this is:

    q1 = sa.select(last_active,integer).select_from(table).subquery()
 
you should not use the .c attribute on a select() object.   should be throwing lots of warnings under 1.4.

if you can show me the SQL you actually want to render, I can show you that.

**Reply by SunilMallireddy** [reply]
Thanks, I meant it o be Thank you for the clarification. My apologies for the mistake.

To rephrase the issue:

The problem is that when using sqlalchemy to create a new datetime column in BigQuery by passing an Interval expression to the date_add function using sa.text(), the query alias for the integer column is incorrect in the subsequent queries. The alias for the last_active column is updated correctly to "anon_2", but the integer column points to "anon_1" instead of "anon_2" in the query that uses the new datetime column. This causes the query to fail. The issue may be due to incorrect usage of the select_from() function in the original code, or some other reason. The question is whether there is an alternative sqlalchemy function that can be used instead of sa.text() to avoid this issue.

Please find the below queries

```python

#create a subquery2
q2 = sa.select(new_col)

#query = 'SELECT date_add(anon_1.last_active, INTERVAL anon_1.integer days) AS newdatetimecolumn \nFROM (SELECT activity.last_active AS last_active, activity.integer AS integer \nFROM activity) AS anon_1'

#create a new query
q3 = sa.select([sa.func.count(sa.text('*'))]).select_from(q2)

#query = str(q3)
##query = 

'SELECT count(*) AS count_1 \nFROM (SELECT date_add(anon_2.last_active, INTERVAL anon_1.integer days) AS newdatetimecolumn \nFROM (SELECT activity.last_active AS last_active, activity.integer AS integer \nFROM activity) AS anon_2) AS anon_1'

```

Expected query:
```
'SELECT count(*) AS count_1 \nFROM (SELECT date_add(anon_2.last_active, INTERVAL anon_2.integer days) AS newdatetimecolumn \nFROM (SELECT activity.last_active AS last_active, activity.integer AS integer \nFROM activity) AS anon_2) AS anon_1'
```

Actual query:

```
'SELECT count(*) AS count_1 \nFROM (SELECT date_add(anon_2.last_active, INTERVAL anon_1.integer days) AS newdatetimecolumn \nFROM (SELECT activity.last_active AS last_active, activity.integer AS integer \nFROM activity) AS anon_2) AS anon_1'

```

**Reply by zzzeek** [reply]
so easiest way to deal with that is to just give your subqueries a specific name:

    anon_2 = sa.select(table.c.last_active, table.c.integer).subquery("anon_2")

now you can refer to that name inside of text() or literal_column() (literal_column is better for column expressions).

However, to render that INTERVAL phrase generically you can also make a [compiles](https://docs.sqlalchemy.org/en/20/core/compiler.html) recipe very simply, see the below proof of concept.

```py
import sqlalchemy as sa
from sqlalchemy.sql import ColumnElement
from sqlalchemy.ext.compiler import compiles

##create a new table
table = sa.table(
    "activity",
    sa.column("last_active", sa.DateTime()),
    sa.column("integer", sa.Integer()),
)

"""
SELECT count(*) AS count_1
FROM (
    SELECT date_add(anon_2.last_active, INTERVAL anon_2.integer days) AS newdatetimecolumn
    FROM (
        SELECT activity.last_active AS last_active, activity.integer AS integer FROM activity
    ) AS anon_2
) AS anon_1'

"""


class INTERVAL(ColumnElement):
    def __init__(self, element, days):
        self.element = element
        self.days = days
        self.type = sa.Interval()


@compiles(INTERVAL)
def _compile_interval(element, compiler, **kw):
    return f"INTERVAL {compiler.process(element.element, **kw)} {element.days}"


anon_2 = sa.select(table.c.last_active, table.c.integer).subquery() # "anon_2")
anon_1 = sa.select(
    sa.func.date_add(
        anon_2.c.last_active,
        INTERVAL(anon_2.c.integer, "days")
    ).label("newdatetimecolumn")
).subquery() # "anon_1")
stmt = sa.select(sa.func.count()).select_from(anon_1)

print(stmt)

```

output:

```sql
SELECT count(*) AS count_1 
FROM (SELECT date_add(anon_2.last_active, INTERVAL anon_2.integer days) AS newdatetimecolumn 
FROM (SELECT activity.last_active AS last_active, activity.integer AS integer 
FROM activity) AS anon_2) AS anon_1
```

**Reply by SunilMallireddy** [reply]
> so easiest way to deal with that is to just give your subqueries a specific name:
> 
> ```
> anon_2 = sa.select(table.c.last_active, table.c.integer).subquery("anon_2")
> ```
> 
> now you can refer to that name inside of text() or literal_column() (literal_column is better for column expressions).
> 
> However, to render that INTERVAL phrase generically you can also make a [compiles](https://docs.sqlalchemy.org/en/20/core/compiler.html) recipe very simply, see the below proof of concept.
> 
> ```python
> import sqlalchemy as sa
> from sqlalchemy.sql import ColumnElement
> from sqlalchemy.ext.compiler import compiles
> 
> ##create a new table
> table = sa.table(
>     "activity",
>     sa.column("last_active", sa.DateTime()),
>     sa.column("integer", sa.Integer()),
> )
> 
> """
> SELECT count(*) AS count_1
> FROM (
>     SELECT date_add(anon_2.last_active, INTERVAL anon_2.integer days) AS newdatetimecolumn
>     FROM (
>         SELECT activity.last_active AS last_active, activity.integer AS integer FROM activity
>     ) AS anon_2
> ) AS anon_1'
> 
> """
> 
> 
> class INTERVAL(ColumnElement):
>     def __init__(self, element, days):
>         self.element = element
>         self.days = days
>         self.type = sa.Interval()
> 
> 
> @compiles(INTERVAL)
> def _compile_interval(element, compiler, **kw):
>     return f"INTERVAL {compiler.process(element.element, **kw)} {element.days}"
> 
> 
> anon_2 = sa.select(table.c.last_active, table.c.integer).subquery() # "anon_2")
> anon_1 = sa.select(
>     sa.func.date_add(
>         anon_2.c.last_active,
>         INTERVAL(anon_2.c.integer, "days")
>     ).label("newdatetimecolumn")
> ).subquery() # "anon_1")
> stmt = sa.select(sa.func.count()).select_from(anon_1)
> 
> print(stmt)
> ```
> 
> output:
> 
> ```sql
> SELECT count(*) AS count_1 
> FROM (SELECT date_add(anon_2.last_active, INTERVAL anon_2.integer days) AS newdatetimecolumn 
> FROM (SELECT activity.last_active AS last_active, activity.integer AS integer 
> FROM activity) AS anon_2) AS anon_1
> ```

Thanks, It's working as expected when I tried with the compile process.
